cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java

```diff
@@ -67,16 +81,54 @@ public BusinessMdsIterable get() {
         }
       }
     });
+    this.cConf = cConf;
+    this.changePublisher = changePublisher;
   }
 
   /**
    * Adds/updates metadata for the specified {@link Id.NamespacedId}.
    */
-  public void setProperties(final Id.NamespacedId entityId, final Map<String, String> metadata) {
+  public void setProperties(final Id.NamespacedId entityId, final Map<String, String> properties) {
+    if (!cConf.getBoolean(Constants.Metadata.UPDATES_PUBLISH_ENABLED)) {
+      setPropertiesNoPublish(entityId, properties);
+      return;
+    }
+    final ImmutableMap.Builder<String, String> propAdditions = ImmutableMap.builder();
+    final ImmutableMap.Builder<String, String> propDeletions = ImmutableMap.builder();
+    final AtomicReference<MetadataRecord> previousRef = new AtomicReference<>();
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        Map<String, String> existingProperties = input.businessMds.getProperties(entityId);
+        Set<String> existingTags = input.businessMds.getTags(entityId);
+        previousRef.set(new MetadataRecord(entityId, existingProperties, existingTags));
+        for (Map.Entry<String, String> entry : properties.entrySet()) {
+          String existingValue = existingProperties.get(entry.getKey());
+          if (existingValue != null && existingValue.equals(entry.getValue())) {
+            // Value already exists and is the same as the value being passed. No update necessary.
+            continue;
+          }
+          // At this point, its either an update of an existing property (1 addition + 1 deletion) or a new property.
+          // If it is an update, then mark a single deletion.
+          if (existingValue != null) {
+            propDeletions.put(entry.getKey(), existingValue);
+          }
+          // In both update or new cases, mark a single addition.
+          propAdditions.put(entry.getKey(), entry.getValue());
+          input.businessMds.setProperty(entityId, entry.getKey(), entry.getValue());
+        }
+        return null;
+      }
+    });
+    publish(previousRef.get(), new MetadataRecord(entityId, propAdditions.build(), EMPTY_TAGS),
+            new MetadataRecord(entityId, propDeletions.build(), EMPTY_TAGS));
+  }
+
+  private void setPropertiesNoPublish(final Id.NamespacedId entityId, final Map<String, String> properties) {
```

Comment:
The main reason for doing that was to avoid the scans to get current state and calculate the diff, when publishing is disabled. Let me think if there is a better way of doing it.